regression-tests/pure2-last-use.cpp2

```diff
@@ -0,0 +1,244 @@
+    f_inout(a);
+    f_copy(this);
+  }
+  p3: (move this) = {
+    f_inout(this);
+    f_copy(a);
+  }
+  q: (move this) = {
+    h();
+    n();
+  }
+// FIXME Lack of implicit moves in sibling selections.
+// Not a `this` problem: <https://cpp2.godbolt.org/z/hbeje4c3M>.
+//z: (move this, move that) = {
+///*f */      if true { f_copy(this); }
+///*f */ else if true { f_copy(this, that); }
+///*g */ else if true { f_copy(this.a); }
+///*g */ else if true { f_copy(this.a, that.a); }
+///*i */ else if true { f_copy(a); }
+///*j */ else if true { f_copy(a); }
+///*k */ else if true { f_copy(a, b); }
+///*l */ else if true { k(); }
+///*m */ else if true { this.k(); }
+///*o1*/ else if true { n(); }
+///*o2*/ else if true { this.n(); }
+///*o3*/ else if true { n(0); }
+///*o4*/ else         { n(this); }
+//}
+}
+
+issue_857_2: @struct type = {
+  a: std::unique_ptr<int>; // OK: No error about 'a' being unused.
+}
+
+gi: int = 0;
+issue_857_3: @struct type = {
+  i: std::add_lvalue_reference_t<int> = gi;
+  f: (move this) = _ = f_inout(i); // OK: The implicit `this` is moved, not `i`.
+}
+
+issue_857_4: @struct type = {
+   f: std::add_pointer_t<int()>;
+   g: std::add_pointer_t<int(int)>;
+  mf: std::add_pointer_t<int()>;    // TODO Use `std::move_only_function`.
+  mg: std::add_pointer_t<int(int)>; // Ditto.
+  h0: (move this) = _ = mf();
+  h1: (move this) = _ = this.mf();
+  h2: (move this, that) = _ = that.mf();
+  h3: (move this, that) = _ = that.f;
+  h4: (move this, x: int) = _ = x.mg();
+  h5: (move this, x: int) = _ = f() + x.g();      // FIXME #313.
+  h6: (move this, x: int) = _ = x.g() + f();      // FIXME #313.
+  h7: (move this, x: int) = _ = this.f() + x.g(); // FIXME #313.
+  h8: (move this, x: int) = _ = x.g() + this.f(); // FIXME #313.
+  i0: (move this, x: int) = {
+    _ = f();
+    _ = x.mg();
+  }
+  i1: (move this, x: int) = {
+    _ = x.g();
+    _ = mf();
+  }
+  i2: (move this, x: int) = {
+    _ = this.f();
+    _ = x.mg();
+  }
+  i3: (move this, x: int) = {
+    _ = x.g();
+    _ = this.mf();
+  }
+}
+
+issue_869: @union type = {
+  i: int;
+}
+
+issue_884_3: () = {
+  x := new<int>(0);
+  if true { }
+  if true { }
+  {
+    { f_inout(x); }
+    f_copy(x);
+  }
+}
+
+issue_884: () = {
+  x := new<int>(0);
+  if true { }
+  {
+    { f_inout(x); }
+    f_copy(x);
+  }
+}
+
+issue_884_2: () = {
+  x := new<int>(0);
+  if true { }
+  {
+    f_inout(x);
+  }
+  f_copy(x);
+}
+
+issue_888_0: (copy r: std::string, copy size: int) = {
+  _ = r.size();
+}
+// TODO Use `std::move_only_function`.
+issue_888_1: (copy _: std::string, copy size: std::add_pointer_t<int(int)>) = {
+  _ = 0.size();
+}
+
+draw: () = {
+  pos       := 0;
+  vertex    := :(_) = { }; // TODO Use `std::move_only_function`.
+  _ = (pos).vertex();
+}
+
+enum_0: () = {
+  underlying_type: int;
+  if true { }
+  underlying_type = 0;
+}
+
+enum_1: () = {
+  underlying_type: int;
+  if true {
+    underlying_type = 0;
+  }
+  underlying_type = 0;
+}
```

Comment:
Executing this results in a contract violation.
Looks like the approach with implicit `else` branches doesn't really work.

Reply:
Right, that would be a classic "variable has to be initialized on both branches or neither, and it's not being initialized on the implicit `else` branch" diagnostic.

Making line 223 be part of an implicit `else` can't work because we have to know whether to emit that `=` as a construction or an assignment (and logically it must be an assignment because it's not a definite first use -- yes I know I need to write the language guide that covers that, I've started writing that but I also don't want to keep getting more behind on issues/PRs -- thanks to everyone for those and for your patience).

Reply:
> I know I need to write the language guide that covers that, I've started writing that but I also don't want to keep getting more behind on issues/PRs

Let me know if you need help in that area.

Reply:
Yes, thank you! I haven't forgotten your kind offer, I just want to get enough down to figure out the structure and approach and then there'll be sections to fill in.